annotator-models/trainer/dawid_skene.py

```diff
@@ -0,0 +1,565 @@
+
+    item_classes = np.zeros([nItems, nClasses])
+
+    for i in range(nItems):
+
+        counts_i = np.stack([counts[i,:,:],counts[i,:,:]],axis=1)
+        estimate_1 = np.prod(np.power(error_rates,counts_i), axis=(0,2))
+        estimate_1 = class_marginals *  estimate_1
+        item_classes[i,:] = estimate_1
+
+        # normalize error rates by dividing by the sum over all classes
+        item_sum = np.sum(item_classes[i,:])
+        if item_sum > 0:
+            item_classes[i,:] = item_classes[i,:]/float(item_sum)
+
+    return item_classes
+
+def e_step_verbose(counts, class_marginals, error_rates):
+    """
+    Determine the probability of each item belonging to each class,
+    given current ML estimates of the parameters from the M-step
+    See equation 2.5 in Dawid-Skene (1979)
+
+    Inputs:
+      counts: Array of how many times each rating was given
+          by each rater for each item
+      class_marginals: probability of a random item belonging to each class
+      error_rates: probability of rater k assigning a item in class j
+          to class l [raters, classes, classes]
+
+    Returns:
+      item_classes: Soft assignments of items to classes
+          [items x classes]
+    """
+    [nItems, nRaters, nClasses] = np.shape(counts)
+
+    item_classes = np.zeros([nItems, nClasses])
+
+    for i in range(nItems):
+        for j in range(nClasses):
+            estimate = class_marginals[j]
+            estimate *= np.prod(np.power(error_rates[:,j,:], counts[i,:,:]))
+
+            item_classes[i,j] = estimate
+
+        # normalize error rates by dividing by the sum over all classes
+        item_sum = np.sum(item_classes[i,:])
+
+        if item_sum == 0 or math.isnan(item_sum):
+            continue
+
+        item_classes[i,:] = item_classes[i,:]/float(item_sum)
+
+    return item_classes
+
+def calc_likelihood(counts, class_marginals, error_rates):
+    """
+    Calculate the likelihood given the current parameter estimates
+    This should go up monotonically as EM proceeds
+    See equation 2.7 in Dawid-Skene (1979)
+
+    Inputs:
+      counts: Array of how many times each response was received
+          by each rater from each item
+      class_marginals: probability of a random item belonging to each class
+      error_rates: probability of rater k assigning a item in class j
+          to class l [raters, classes, classes]
+
+    Returns:
+      Likelihood given current parameter estimates
+    """
+    [nItems, nRaters, nClasses] = np.shape(counts)
+    log_L = 0.0
+
+    for i in range(nItems):
+        item_likelihood = 0.0
+        for j in range(nClasses):
+
+            class_prior = class_marginals[j]
+            item_class_likelihood = np.prod(np.power(error_rates[:,j,:], counts[i,:,:]))
+            item_class_posterior = class_prior * item_class_likelihood
+            item_likelihood += item_class_posterior
+
+        temp = log_L + np.log(item_likelihood)
```

Comment:
This function is written such that the program exits if the likelihood for any example is ever infinity or nan since that should never happen. The equation for the log likelihood function includes an outer sum over all items (the likelihood of the data is the product over the likelihood of each example, the log likelihood is the sum over all examples). We store the likelihood for each example in a `temp` variable, check that it's not infinity or nan and if not, add it to the running sum, ` log_L`.